In [1]:
import pandas as pd

In [10]:
from google.colab import files
import os

# Check if the file exists, if not, prompt for upload
file_name = 'Titanic-Dataset.csv'
if not os.path.exists(file_name):
  uploaded = files.upload()
  if file_name not in uploaded:
    print(f"Warning: '{file_name}' not uploaded. Please upload the file.")
  else:
    print(f"'{file_name}' uploaded successfully.")

try:
  titanic_data = pd.read_csv(file_name)
  print("Dataset loaded successfully.")
  print(titanic_data.head())
  print('\n')

  print(titanic_data.info())
  print('\n')

  print(titanic_data.describe())

except FileNotFoundError:
  print(f"Error: The file '{file_name}' was not found. Please ensure it is uploaded.")
except Exception as e:
  print(f"An error occurred: {e}")

Dataset loaded successfully.
   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450

In [14]:
# Data Cleaning

titanic_data['Age'] = titanic_data['Age'].fillna(titanic_data['Age'].median())

embarked_mode = titanic_data["Embarked"].mode()
if not embarked_mode.empty:
    titanic_data["Embarked"] = titanic_data["Embarked"].fillna(embarked_mode[0])
else:
    titanic_data["Embarked"] = titanic_data["Embarked"].fillna('S')

if 'Cabin' in titanic_data.columns:
    titanic_data.drop(columns=['Cabin'], inplace=True)

print(titanic_data.isnull().sum())

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


# Data Analysis

In [17]:
survival_gender = titanic_data.groupby('Sex')['Survived'].mean()
survival_class = titanic_data.groupby('Pclass')['Survived'].mean()
avg_age_per_class = titanic_data.groupby('Pclass')['Age'].mean()

print(survival_gender, '\n')
print(survival_class, '\n')
print(avg_age_per_class, '\n')

# Age Groups
bins = [0, 12, 18, 35, 60, 100]
labels = ['Child', 'Teenager', 'Adult', 'Middle-aged', 'Senior']
titanic_data['AgeGroup'] = pd.cut(titanic_data['Age'], bins=bins, labels=labels)

age_group_survival = titanic_data.groupby('AgeGroup', observed=True)['Survived'].mean()
print(age_group_survival)

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64 

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64 

Pclass
1    36.812130
2    29.765380
3    25.932627
Name: Age, dtype: float64 

AgeGroup
Child          0.579710
Teenager       0.428571
Adult          0.353271
Middle-aged    0.400000
Senior         0.227273
Name: Survived, dtype: float64


# Filtering

In [22]:
# female servivors
female_servivors = titanic_data[(titanic_data['Sex']=='female') & (titanic_data['Survived']==1)]
print(female_servivors)


     PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
874          875         1       2   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ... 

In [23]:
# chlidren servivors
children_servivors = titanic_data[(titanic_data['Age'] < 18 ) & (titanic_data['Survived']==1)]
print(children_servivors)


     PassengerId  Survived  Pclass                                      Name  \
9             10         1       2       Nasser, Mrs. Nicholas (Adele Achem)   
10            11         1       3           Sandstrom, Miss. Marguerite Rut   
22            23         1       3               McGowan, Miss. Anna "Annie"   
39            40         1       3               Nicola-Yarred, Miss. Jamila   
43            44         1       2  Laroche, Miss. Simonne Marie Anne Andree   
..           ...       ...     ...                                       ...   
830          831         1       3   Yasbeck, Mrs. Antoni (Selini Alexander)   
831          832         1       2           Richards, Master. George Sibley   
853          854         1       1                 Lines, Miss. Mary Conover   
869          870         1       3           Johnson, Master. Harold Theodor   
875          876         1       3          Najib, Miss. Adele Kiamie "Jane"   

        Sex    Age  SibSp  Parch       

In [24]:

# first class servivors
first_class_servivors = titanic_data[(titanic_data['Pclass'] == 1) & (titanic_data['Survived']==1)]
print(first_class_servivors)

     PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
11            12         1       1   
23            24         1       1   
31            32         1       1   
..           ...       ...     ...   
862          863         1       1   
871          872         1       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female  28.0      1   
..                                                 ...     ...   ... 

# Insights


1. Who was more likely to survive? (Gender)

Females were significantly more likely to survive.

* Female survival rate: ≈74.2%

* Male survival rate: ≈18.9%

The "women and children first" protocol is clearly visible in these numbers.


2. Did class affect survival?

Yes, there was a direct correlation between social class and survival probability. Passengers in higher classes had much better survival rates:

* 1st Class: ≈63.0%

* 2nd Class: ≈47.3%

* 3rd Class: ≈24.2%

# 4. What combination had the highest survival rate?

The combination with the absolute highest survival rate (100%) was:

* Female children in 2nd Class * Male children in 1st Class

Following closely behind were Adult females in 1st Class, who had a survival rate of approximately 97.4%

In [29]:
print(titanic_data.groupby('Sex')['Survived'].mean() * 100)
print(titanic_data.groupby('Pclass')['Survived'].mean() * 100)
print(titanic_data[titanic_data['AgeGroup'] != 'Unknown'].groupby('AgeGroup')['Survived'].mean() * 100)
print(titanic_data[titanic_data['AgeGroup'] != 'Unknown'].groupby(['Sex', 'Pclass', 'AgeGroup'])['Survived'].mean().sort_values(ascending=False) * 100)

Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64
Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64
AgeGroup
Adult    36.118252
Child    53.982301
Name: Survived, dtype: float64
Sex     Pclass  AgeGroup
female  2       Child       100.000000
male    1       Child       100.000000
female  1       Adult        97.674419
        2       Adult        90.625000
        1       Child        87.500000
male    2       Child        81.818182
female  3       Child        54.285714
                Adult        48.623853
male    1       Adult        34.745763
        3       Child        23.255814
                Adult        12.171053
        2       Adult         8.247423
Name: Survived, dtype: float64
